### PR TITLE
Fix SI-5937.

### PR DESCRIPTION
**Pull request by axel22**
Fixes SI-5937 by checking the type of the builder factory and delegating to a special implementation in the case that the builder factory is constructing vectors.

### CODE REVIEW

File: src/library/scala/collection/immutable/Vector.scala [hunk 1]
```diff
@@ -18,8 +18,14 @@ import scala.collection.parallel.immutable.ParVector
 /** Companion object to the Vector class
  */
 object Vector extends SeqFactory[Vector] {
+  private[collection] class VectorReusableCBF extends GenericCanBuildFrom[Nothing] {
+    override def apply() = newBuilder[Nothing]
+  }
+  
+  lazy val VectorReusableCBF: GenericCanBuildFrom[Nothing] = new VectorReusableCBF
```
Comment (retronym) [inline src/library/scala/collection/immutable/Vector.scala]: Does this need to be public?

Comment (jsuereth) [inline src/library/scala/collection/immutable/Vector.scala]: +1.  Also, does it need to be lazy?  Doesn't appear to suck up much space, and object are already lazily initialized.

i'd rather not take the synchronized/voltile hit for this guy in general code.   I think we don't need lazy so much.

**Comment by jsuereth** [comment]
Good fix!  I just think we should remove the `lazy` from the `val` and perhaps make it private.  AFTER THAT, LGTM, even if you didn't list a reviewer :)

**Comment by adriaanm** [comment]
yes, sneakily flying under the radar, huh @axel22? :)
good thing Josh's stealth detectors were turned up to 11

File: src/library/scala/collection/immutable/Vector.scala (hunk 1 above)
Comment (axel22) [inline src/library/scala/collection/immutable/Vector.scala]: Lazy - my first impression is it does not have to be, but a similar thing is lazy in GenTraversableFactory, apparently due to: https://issues.scala-lang.org/browse/SI-4789 (I haven't cast  a detailed look yet).

Public - no probably not.

**Comment by axel22** [comment]
Sorry, was a bit in a rush today to cover everything, so forgot to list you as a reviewer :)
I'll be sure to include you in a couple of other pull requests to make it up ;)

**Comment by axel22** [comment]
Alright, it seems that the issue with the `lazy val` is fixed now, so we should be able to remove the `lazy` on `ReusableCBF` in `GenTraversableFactory` as well.

**Comment by jsuereth** [comment]
Awesome! LGTM.

**Pull request merged**